### ISSUE TITLE
Update several properties of a single object

### ISSUE THREAD

**Issue by JeanCollas** [dotnet/csharplang#1449]
I don't find any way to update several properties of a single object in C# (similar to the with VB statement). 
It would be useful to clarify some mapping functions where several properties has to be updated at once.
Here is a very short example

        public class MyLink
        {
            public int Id { get; set; }
            public string Name { get; set; }
            public string Description { get; set; }
            public string Link { get; set; }
            public MetaData Meta { get; set; }
            public DateTime? Deleted { get; set; }
        }
        public class MetaData
        {
            public DateTime LastUpdate { get; set; }
            public decimal AvgRate { get; set; }
        }

        public void UpdateLink(MyLink link, string name, string description, DateTime lastUpdate)
        {
            link.{
                Name = name,
                Description = description,
                Meta.{
                   LastUpdate = lastUpdate
                }
            };
        }

Although it might induce some recursive conflicts.

**Comment by jnm2** [comment]
See https://github.com/dotnet/csharplang/issues/803 too.

**Comment by jnm2** [comment]
See https://github.com/dotnet/csharplang/issues/77 and https://github.com/dotnet/csharplang/issues/803 too.

**Comment by JeanCollas** [comment]
Indeed, not easy to find if there is a relevant issue corresponding :) 
However I prefer the use of a '.' than a keyword which does not look like C# syntax. As mentioned in #1434 it can be followed by a question mark '.?' to avoid null exceptions.

**Comment by jnm2** [comment]
@JeanCollas Certainly not easy! No problem.

C# is adding many new keywords such as `catch (Exception ex) when (ex.X == 42)` and `case Foo foo when foo.X == 42` and besides `x is null`, there are more proposed for pattern matching such as `and`, `or` and `not`. Not to mention LINQ. A `with` keyword seems consistent with those, doesn't it?

**Comment by jnm2** [comment]
@JeanCollas Certainly not easy! No problem.

C# is adding many new keywords such as `catch (Exception ex) when (ex.X == 42)` and `case Foo foo when foo.X == 42` and besides `x is null`, there are more proposed for pattern matching such as `and`, `or` and `not`. Not to mention LINQ. A `with` keyword seems consistent with those, doesn't it?

Now whether `.` could be a better syntax than `when`, maybe? I'm not fond of `foo.{` which feels a little like we're heading toward ascii sneeze, but it's really not that bad and you make good points.

**Comment by JeanCollas** [comment]
@jnm2 

For what I have in mind, that kind of keywords in only filters/test, a bit like linq or SQL syntax, or could be declarative like 'if(oldVar is Type myVar)'.

Except for this case, not so many keywords are used for assignment shortcut/value handling, it is most of the time done by operators ( =>, autoproperties {get;set;} = defaultValue, ...)
That is the reason why I don't see it as a great improvement. 

However I will still be fine of course if this is the final choice of the team, but I am afraid that C# gets too verbose/complicated instead of keeping it straight (like javascript syntax). The shorter is the code, the easiest to read it is, and keep the words for what really needs explanations (how is the method, how is the class, etc).

To focus on the "with" keyword, I feel it would also be confusing as "does it returns the given object with updated values" or "does it gives a new object, copied from the given object, but also with the new values".

If you take a List<string> for example, I would feel like
list.With("A", "B") would give me a new list with merged content, and not the initial list with two new values.

Actually to be more precise, I would not expect a new list with merged content, nor an update of the list, but a VIEW of the original list, with these two additional values, like a new presenter of the object with overridden properties.

**Comment by JeanCollas** [comment]
From what I see there: https://github.com/DavidArno/SuccincT/issues/44 it is already planned as ' with ' syntax for C#8. So I guess any discussion is now just almost vain :D

**Comment by DavidArno** [comment]
@JeanCollas,

Please don't go trusting statements made by random folk on the internet, especially when that random folk is me! 😀 

[`with` syntax is currently assigned to the C# 8.0 bucket](https://github.com/dotnet/csharplang/issues?q=is%3Aopen+is%3Aissue+milestone%3A%228.0+candidate%22). But there is no guarantee that it will make it into that release and that release may still be a way off (7.3 is the next release, likely due in the next couple of months). So I'd say that the discussion is potentially far from over.

**Comment by HaloFour** [comment]
Isn't `with` about creating a new instance with the same property values except for those explicitly specified in the initializer anyway?  It's not the same as trying to use initializer-like syntax to modify the current instance.

There are some other proposals around supporting initializer syntax on expressions to support factory methods.  That would probably be very similar to this proposal, sans the dot operator.  But the team have also expressed little interest in the various proposals to promote mutation of the current state of an instance so I'm not convinced that such proposals have much in the way of legs at the moment.

**Comment by JeanCollas** [comment]
@DavidArno I will take this advice seriously :D 
I am not fan of the `with` syntax solution, and I hope it will be improved.

@HaloFour It is actually not really clear to me. After more reading, I guess you are right. In that case, I believe this `with` operator could lead to potential unexpected forbidden states, and shall be used carefully.

**Comment by Starwer** [comment]
It would be nice if this could work to initializers, and to methods.
I think that a WinForm code could eventually look structured with this proposal.

**Comment by DavidArno** [comment]
@JeanCollas,

I see MS proved my point about trying to guess releases of the language. [On the very same day that I suggest C# 7.3 might appear in a month or two, they release C# 7.3](https://docs.microsoft.com/en-us/visualstudio/releasenotes/vs2017-Preview-relnotes#15.7.0%20Preview%203)! 😆

**Comment by Makeman-from-Makeloft** [comment]
@JeanCollas, hello!

> I don't find any way to update several properties of a single object in C# (similar to the with VB statement).
> It may be useful to clarify some mapping functions where several properties has to be updated at once. 

There is the simple but not so obvious way since C# 7 (and partially from earlier versions).

_Note:_ possible, this comment will be downvoted or even deleted by my lovely haters ❤ but I feel that should share the approach with you too.

At first, consider two extension methods
```c#
public static T To<T>(this T o, out T x) => x = o;
public static T With<T>(this T o, params object[] pattern) => o;
```
This methods allows to rewrite your code like
```c#
public void UpdateLink(MyLink link, string name, string description, DateTime lastUpdate) =>
    link.To(out var l)?.With(
        l.Name = name,
        l.Description = description,
        l.Meta.To(out var m)?.With(
           m.LastUpdate = lastUpdate
        )
    );
```
So, as you see it is a little bit more glibly but fully workable solution [[an online sample](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAIAYAEqCMBuAWAChUBmbHANmwCYKB2TAb2Mzc1fbNswFkBPADIBLKAGtObFkXazs5UcEwBJOM0wBzAKbA8mAM469AX0lzuuLADkAhgFst67boNHMpmXLYWcWACJa+gDGYMIADsDCAPZQTm6GLh5e3uSWmCLicS4JJmay3Lw6Nn42wDZ8RVl6Oe55XOQlwFoAKsIOAPyYAQA2OlpqTJrxbknso2xmFnSFZY02ZtLJ8l2lLW2Ogjb6wACqYXCrVa6JdSmYcFpBbTbdmACCAG4aAEqHg87VI2Yek6lU2AAWTB7A5NDJiAAUwAAFsJ9HwhKIxJhukiYBQsFB7Fp0WkLsFQhForAVk1Wg4UVtdvtVgBKTAAXgAfKcUUiAHTNKIQqIQJQPGxgFG09rsgDqwhhENZsm67NsFIZmCxDhgMvYcoCBPCkRijPOgRCOuJas8SzYcpmNk53N5/MFmDsIvFkuh0rN5vYdnZm22IMOSu6VP9TXVbFprNphA9vwo1FQQNwAA4IfTFskoFoAO4I8Gpm08vmYAVC1HiWkuqVlsTsq36zM5q1zVMRj1LdkhrR5gBEKq06G76O7+JCA/RjTWDnlUSzUdZrNwAE4IdX5di522uDhl6utUaiTEN0slyuOVbRb7qaCtEfkqzVxKq2fKkqoBBut1W562B2aWCkRCvbYjgg6YMOhpgCB46rOSWjTrO0ZLAu26nuIa4OLeXgnruEHGoeiHJNhz5lBewZ/jeBFsB4PwkH88Z0IIACixDpmcuDUM0mBcgAPM0TJQrC8KcVE6J2lxmAAB70sykn6lE0YAJA+BxmCPtCvH8TCcLiSJmBhIK9jwlEABGABWlzAAA2gAunppRNGAUDSUymDycQCkKe5yniWpGkCdpwnOa5ineZxvnNOidyaYJOmRZgNhBW5RBKXRPmurxkXogAQtFAWuXFNjZZgxmJSFqVhelEX3Nl6IAMK5UJ+X3PFRXGXVmBBKVZgKQA9AAVJg7JDZgfU9cQHhAA==)] which support recursive expressions and null propagation operator.

But the suggested implementation of `With` method has some lacks
- array allocations
- possible boxing for value types

This problem may be solved by this way
```c#
public static T With<T>(this T o) => o;
public static T With<T, A>(this T o, A a) => o;
public static T With<T, A, B>(this T o, A a, B b) => o;
public static T With<T, A, B, C>(this T o, A a, B b, C c) => o;
    	    	/* ... */
```
If you want to a get very-very long performance optimized `With` statement it is possible to concatenate (to glue) shorter statements by several times
```c#
GetModel().To(out var m).With(m.A0 = a0, ... , m.AN = an).With(m.B0 = b0, ... ,m.BM = bM).Save();
```
This approach has very good performance extremely close to ideal.

There is a side effect related with structs [value types]. For example if you want modify and then return modified struct into a method calls chain you should apply following extensions
```c#
public static TR Like<T, TR>(this T o, TR y) => y;
public static TR Like<T, TR>(this T o, ref TR y) => y;
```
Because a struct will be copied during a method call and it will be returned the original copy instead an already modified instance
```c#
static AnyStruct SetDefaults(this AnyStruct s) => s.With(s.Name = "DefaultName").Like(ref s);
```
New versions of C# support `this ref` for extension methods but you should to use another name of `With` method to avoid ambiguously of call
```c#
public static T WithRef<T, A>(this ref T o, A a) where T: struct => o;
```
Also `With` method useful too into similar scenarios
```c#
// possible NRE
void UpdateAppTitle() => Application.Current.MainWindow.Title = title;

// currently not supported by C#, possible, will be added later
void UpdateAppTitle() => Application.Current.MainWindow?.Title = title;

// classical solution
void UpdateAppTitle() {
    var window = Application.Current.MainWindow;
    if (window != null) window.Title = title;
}

void UpdateAppTitle() => Application.Current.MainWindow.To(out var w)?.With(w.Title = title);
```

This is basic technical information about `to-with` pattern in C# but not full.

Consider closely some foundation questions. There are two directions (kinds) of assignment
```c#
IModel m;
m = GetModel(); // left side assignment
GetModel().To(out m); // right side assignment
```
Yes, all functions with `out` and partially with `ref` parameters are just kinds of right side assignment.

From earlier versions C# have support `out & ref` parameters. There was some benefits but not very impressive. But C# 7 made an evolutionary leap!

A small syntax sugar feature `o.To(out var x)` allows to use right side variable declaration and assignment together which allows to use by a convenient way `To-With` pattern, generalize and clarify some common scenarios in programming...

Historically there is much more habitually classical left side assignment. Possible it is influence of mathematics where `y = f(x)` is a standard notation. But in the practice of programming this cause some difficulties for example 'parentheses hell' with chained type castings
```c#
public void EventHandler(object sender, EventArgs args) =>
    ((IModel) ((Button) sender).DataContext).UpdateModel();

// in a general case there is not possible settle priorities without parentheses
// (IModel) (Button) sender.DataContext.UpdateModel();
```
It forces some people to use verbose or bad solutions like
```c#
/* NullReferenceException instead of InvalidCastException */
public void EventHandler(object sender, EventArgs args) =>
    ((sender as Button).DataContext as IModel).UpdateModel();

/* miss of InvalidCastException */
public void EventHandler(object sender, EventArgs args) =>
    ((sender as Button)?.DataContext as IModel)?.UpdateModel();

/* verbose */
public void EventHandler(object sender, EventArgs args)
{
    var button = (Button) sender;
    var model = (IModel) button.DataContext;
    model.UpdateModel();
}
```
But this problem may be solved in a much more elegant manner
```c#
public void EventHandler(object sender, EventArgs args) =>
    sender.To<Button>().DataContext.To<IModel>().UpdateModel();
    
public static T To<T>(this object o) => (T) o;
```
Then if we generalize the approach we get following set of extension methods
```c#
private static object ChangeType(this object o, Type type) =>
    o == null || type.IsValueType || o is IConvertible ? Convert.ChangeType(o, type, null) : o;

public static T To<T>(this T o) => o;
public static T To<T>(this T o, out T x) => x = o;
public static T To<T>(this object o) => (T) ChangeType(o, typeof(T));
public static T To<T>(this object o, out T x) => x = (T) ChangeType(o, typeof(T));
```
We able to use too
```c#
sender.To(out Button b).DataContext.To(out IModel m).UpdateModel();
/* or */
sender.To(out Button _).DataContext.To(out IModel _).UpdateModel();
```
This samples illustrate that historically C# missed something like an operator `to`.
Compare
```c#
((sender to Button b).DataContext to IModel m).UpdateModel();
((sender to Button _).DataContext to IModel _).UpdateModel();
/* or even */
sender to Button b.DataContext to IModel m.UpdateModel();
sender to Button _.DataContext to IModel _.UpdateModel();
```
There are some interesting facts about C# operators `as & is` but it is a theme for another spoke.

Back to `To-With` pattern.
Not obvious but very important thing that pattern is both oriented [duplex, in-out, right-left, mirror symmetrically]. What is mean?

It is mean that 
```c#
GetPerson().To(out var p).With
(
        /* deconstruction-like variations */
	p.Name.To(out var name), /* right side assignment to the new variable */
	p.Name.To(out nameLocal), /* right side assignment to the declared variable */
	NameField = p.Name, /* left side assignment to the declared variable */
	NameProperty = p.Name, /* left side assignment to the property */

        /* a classical initialization-like variation */
	p.Name = "AnyName"
)
```
we can use it for both initialization and deconstruction of objects at the same time!

As we see regular `json`-like object initializers are only a limited particular syntax variation of much more general `with` pattern.

Moreover a similar approach we can apply for collection initializers too and use
```c#
public CustomCollection GetSampleCollection() =>
    new CustomCollection().To(out var c).With(c.Name = "Sample").Merge(a, b, c, d);

/* currently not possible */
public CustomCollection GetSampleCollection() =>
    new CustomCollection { Name = "Sample" } { a, b, c, d };
```

There is possible to implement `check` pattern closely to `with`
```c#
if (GetPerson() is Person p && p.Check(
        p.FirstName is "Keanu",
        p.LastName is string lastName,
        p.Age.To(out var age) > 23
    ).All(true)) ...

case Person p when p.Check(
        p.FirstName.StartWith("K"),
        p.LastName.StartWith("R"),
        p.Age.To(out var age) > 23
    ).Any(true): ...

case Point p when p.Check(
        p.X > 9,
        p.Y > 7 && p.Y < 221
        p.Z > p.Y
        p.T > 0
    ).Count(false) == 2: ...
```
Look at
```c#
public static bool[] Check<T>(this T o, params bool[] pattern) => pattern;
```
More details and implementations of some other interesting things like
```c#
public static double CalculateSquare(this Shape shape) =>
	shape.Match(
		(Line _) => 0,
		(Circle c) => Math.PI * c.Radius * c.Radius,
		(Rectangle r) => r.Width * r.Height,
		() => double.NaN
	);
```
```c#
var words = New.Array("hello", "wonderful", "world");
var ints = New.List(1, 2, 3, 4, 5);
```
```c#
var (x, y, z) = 0;
(x, y, z) = 1;
```
 you may found by links
Github mirror: [implementation](https://github.com/Makeman-from-Makeloft/Ace/tree/master/Ace.Base/Sugar) / [some tests](https://github.com/Makeman-from-Makeloft/Ace/tree/master/Ace.Tests/Ace.Base.Sandbox)
Bitbucket mirror: [implementation](https://bitbucket.org/Makeloft/ace/src/831f3081b07d899f260dc3930d674b60e6095148/Ace.Base/Sugar/?at=master) / [some tests](https://bitbucket.org/Makeloft/ace/src/831f3081b07d899f260dc3930d674b60e6095148/Ace.Tests/Ace.Base.Sandbox/?at=master)

Of course nice to have similar `to-with-check` syntax sugar on a native language level like a reference point for other developers, waymark to good design. But not all people understand me and perceive seriously current proposals.

If you really feel a flavor of provided concepts, relish them in practice!

**Comment by JeanCollas** [comment]
@Makeman-from-Makeloft Thanks for your message. I am not sure this exactly answers my issue with is more about syntax/code structure, but it is very interesting on patterns, and several new points I do not master enough yet. The Check/Match pattern is very nice, and I will probably exploit it :)

**Comment by JeanCollas** [comment]
After a second read of this long post, I only half agree with this. It looks nice, but actually it induces some things similar to global variables / inline delegates that you actually cannot easily extract from the code. 
The possibilities are nice on a first approach, but a deeper reflection will make me not use them because it will look like quick and dirty solutions :/

**Comment by mkmn-xyz** [comment]
posted by @Makeman-from-Makeloft 

@JeanCollas 

As I notified before, my dear haters remove my detailed comment. It is very predictable.

At first, don't believe to me, second, don't believe blindly even to authority members from this repo.
Try to look for common sense.

A backup of the deleted comment is [here](https://bitbucket.org/Makeloft/ace/issues/2/to-with).

Consider your critics :)

`To-With-Check` pattern don't contains global variables or inline delegates. Yes, it implies declaration of local variables by an explicit way. Of course with syntax sugar like

```c#
GetPerson()?.With
{
    .Name = "Makeman",
    .Age = 28
}

/* vs */

GetPerson().To(out var p)?.With
{
    p.Name = "Makeman",
    p.Age = 28
}
```
we can use an implicit local variable closed into a context of `With` expression.

Obviously, even if we use a classical property block initializer we declare a new local variable (see [a decompile](https://sharplab.io/#v2:CYLg1APgAgzABFAjANgQJgYg7HA3gWACg4SF4oMAFAUwCcBnAewDsjS833TZMAGOAHIBDALbU8cAObUALgG449WXIC+nEmuKl1ZTKigAWTAA4AFAEoOWrnABuQ2orgBeOM2oB3ODQYs8wsRc4ACIAWSEAa2oRIWZglQA6ABVGAGUZWgBLZkkLOR1NFSA)).


If you mean that inline delegates used in the following sample
```c#
public static double CalculateSquare(this Shape shape) =>
    shape.Match(
        (Line _) => 0,
        (Circle c) => Math.PI * c.Radius * c.Radius,
        (Rectangle r) => r.Width * r.Height,
        () => double.NaN
    );
```
I want to say that in the majority of real cases it is not cause value performance lacks but brings aesthetic pleasure from beautiful code.

And about dirty... Do you consider closely difference in following expressions?

```c#
Person GetPerson() => ...

GetPerson() is Person p
GetPerson() is var p
```
There is a small dirty spot already shipped into C# 7. But do you listen something about planned `is {...}`?

```c#
GetPerson() is { } p
GetPerson() is { Name is "Makeman" } p
```
Sorry, but it is a clumsy syntax sugar instead using of powerful `With & Check` patterns.

And in conclusion you may think about questions. Why some people constantly try to lock me and remove really detailed comments and workable solutions? If proposals are so wrong why instead argumented responses they prefer to ban the person? Maybe this person shows something true but not very pleasant for them?

P.S. If you feel the community suppression you may directly contact with me [makeman@tut.by]

**Comment by jnm2** [comment]
For those who weren't around earlier: The banning has nothing to do with the workarounds he is suggesting. Personally, like @JeanCollas, we've seen (and tried to discuss) some drawbacks with them, but we are happy for anyone to look at them and use them if they like them.
The reason the community would rather ignore Makeman/Makeloft is due to this person's blatant disregard for the [code of conduct](https://www.dotnetfoundation.org/code-of-conduct). This person is responsible for a series of both in-your-face rude comments and incessant spamming on threads that were not related, and he was recalcitrant to appeals. After many community members complained, the admins banned this person. He has since created dozens of sock puppet accounts and continued the spamming. In other words, all the signs say that talking to this person is a waste of time. 😟

**Comment by FurkanKambay** [comment]
Is there a place for a word like `populate` in that syntax or can it be modified so? I think it's worth considering. But I like the idea and the syntax.

**Comment by AustinBryan** [comment]
Isn't this sort've similar to a `with` keyword in other languages?

```csharp
public class MyLink
{
    public string Name        { get; set; }
    public string Description { get; set; }
    public string Link        { get; set; }
    public MetaData Meta      { get; set; }

    public void DrawLink() { }
}
...
public void UpdateLink(MyLink link, string name, string description, DateTime lastUpdate)
{
    with link
    {
        Name = name;
        Description = description;

        with Meta 
        {
            LastUpdate = lastUpdate;
        }

        DrawLink();
    }
}
```

Basically, the `with` keyword creates a block of code that behaves as if it is a part of that class, allowing you to access the fields and methods directly. It probably shouldn't expose `private` fields and methods, though. 

Regardless of if it uses `.` or `with`, might as well allow the usage of methods inside, right?

**Comment by FurkanKambay** [comment]
@AustinBryan I agree. Also, just a thought: instead of `with`, how about `scope`? Like `scope (link) { }`. Also, I think the brackets are better suited for the language.

**Comment by JeanCollas** [comment]
@AustinBryan I agree this is similar to `with`, however having keywords in the middle of the code like this is not much like C#. Except maybe very last updates, there is no keywords inside the methods that modify the scope (using does, but the usage is quite clear). 
My belief is that it shouldn't be like with or scope, because it induces some interferences if you can add code inside this scope, and how do you manage the following:

    public string Name = "alpha";
    public void T()
    {
        var s=new S() { Name = "beta" };
        with(s)
        {
            Name = "gamma";
        }
        Console.WriteLine(this.Name);
        Console.WriteLine(s.Name);
    }

That is why I think this should not be treated as scope, but only as shortcut

    public void UpdateLink(MyLink link, string name, string description, DateTime lastUpdate)
    {
        link.{
            Name = name,
            Description = description,
            Meta.{
               LastUpdate = lastUpdate
            }
        };
    }

would be translated:

    public void UpdateLink(MyLink link, string name, string description, DateTime lastUpdate)
    {
        link.Name = name;
        link.Description = description;
        link.Meta.LastUpdate = lastUpdate;
    }

which seems quite obvious.

**Comment by AustinBryan** [comment]
@JeanCollas I agree. `with` feels way more high level and I really didn't like that language. (GML, had so many issues, like everything being implicitly an `int`). 

So a few questions, would you be able to call methods too, from within the block? I would think so. And would `private`/`protected` fields be visible? I'd say no. 

What I like about this syntax is that it behaves consistently with `if` braces. Whenever I see:

```csharp
if (b) 
{
    // do stuff
}
``` 

I think, "just make it simpler and do this:

```csharp
if (b)
    // do stuff
```

And in your example, the braces for `Meta` are unnecessary, I instinctively thought to remove them, which makes:

```csharp
// braces
Meta.{
    LastUpdate = lastUpdate;
}

// simplified
Meta.LastUpdate = lastUpdate;
```

Without the braces it's just simply the original syntax, which I really like. If we used `with` or something, we'd have:

```csharp
with (Meta) {
    LastUpdate = lastUpdate;
}

// simplified
with (Meta) LastUpdate = lastUpdate;
```

And that does the exact same thing as the normal way, so there's no point in there being more than one way. That just confuses people.
